tests/actor/test_actor_reentrancy.py

```diff
@@ -0,0 +1,223 @@
+# -*- coding: utf-8 -*-
+
+"""
+Copyright (c) Microsoft Corporation and Dapr Contributors.
+Licensed under the MIT License.
+"""
+
+import unittest
+import asyncio
+
+from unittest import mock
+
+from dapr.actor.runtime.runtime import ActorRuntime
+from dapr.actor.runtime.config import ActorRuntimeConfig, ActorReentrancyConfig
+from dapr.serializers import DefaultJSONSerializer
+
+from tests.actor.fake_actor_classes import (
+    FakeReentrantActor,
+    FakeMultiInterfacesActor,
+    FakeSlowReentrantActor,
+)
+
+from tests.actor.utils import _run
+
+
+class ActorRuntimeTests(unittest.TestCase):
+    def setUp(self):
+        ActorRuntime._actor_managers = {}
+        ActorRuntime.set_actor_config(
+            ActorRuntimeConfig(reentrancy=ActorReentrancyConfig(enabled=True)))
+        self._serializer = DefaultJSONSerializer()
+        _run(ActorRuntime.register_actor(FakeReentrantActor))
+        _run(ActorRuntime.register_actor(FakeSlowReentrantActor))
+        _run(ActorRuntime.register_actor(FakeMultiInterfacesActor))
+
+    def test_reentrant_dispatch(self):
+        _run(ActorRuntime.register_actor(FakeMultiInterfacesActor))
+
+        request_body = {
+            "message": "hello dapr",
+        }
+
+        reentrancy_id = "0faa4c8b-f53a-4dff-9a9d-c50205035085"
+
+        test_request_body = self._serializer.serialize(request_body)
+        response = _run(ActorRuntime.dispatch(
+            FakeMultiInterfacesActor.__name__, 'test-id',
+            "ReentrantMethod", test_request_body, reentrancy_id=reentrancy_id))
+
+        self.assertEqual(b'"hello dapr"', response)
+
+        _run(ActorRuntime.deactivate(FakeMultiInterfacesActor.__name__, 'test-id'))
+
+        # Ensure test-id is deactivated
+        with self.assertRaises(ValueError):
+            _run(ActorRuntime.deactivate(FakeMultiInterfacesActor.__name__, 'test-id'))
+
+    def test_interleaved_reentrant_actor_dispatch(self):
+        _run(ActorRuntime.register_actor(FakeReentrantActor))
+        _run(ActorRuntime.register_actor(FakeSlowReentrantActor))
+
+        request_body = self._serializer.serialize({
+            "message": "Normal",
+        })
+
+        normal_reentrancy_id = "f6319f23-dc0a-4880-90d9-87b23c19c20a"
+        slow_reentrancy_id = "b1653a2f-fe54-4514-8197-98b52d156454"
+
+        async def dispatchReentrantCall(actorName: str, method: str, reentrancy_id: str):
+            return await ActorRuntime.dispatch(
+                actorName, 'test-id', method, request_body, reentrancy_id=reentrancy_id)
+
+        async def run_parallel_actors():
+            slow = dispatchReentrantCall(
+                FakeSlowReentrantActor.__name__, "ReentrantMethod", slow_reentrancy_id)
+            normal = dispatchReentrantCall(
+                FakeReentrantActor.__name__, "ReentrantMethod", normal_reentrancy_id)
```

Comment:
Should this be the passthrough method? That's the only time the reentrant ID is fetched from the ContextVar.

Reply:
I don't understand this comment. Could you explain @halspang?

This is just a bunch of Python weirdness to run two async methods truly in parallel and gather the results to force the interleaving of the processing. It is a sanity check to make sure the ContextVars truly work as we expected, that is, ContextVars are unique to each AsyncTask and are always copied forward into nested AsyncTasks.

I understand this looks weird - @wcs1only and I paired on this.

Reply:
I'm fine with the python itself, my comment was more on how we'd handle the ContextVar in the code. Right now, this path only exercises the setting of the reentrant ID. If we use the passthrough method, it should also have to fetch the reentrant ID from the ContextVar. That's the path I'd like to check with interleaving.

Reply:
@berndverst what do you think about my previous comment?